### PR TITLE
Google mail

### PR DESCRIPTION
**Pull request by MatteiT**

### CODE REVIEW

**Review by Newaluigi — DISMISSED** [review]

File: src/app/pages/google-email/google-email.component.html [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    <button (click)="sortByLabel('sent')">Sent</button>
+    <button (click)="sortByLabel('drafts')">Drafts</button>
+    <button (click)="sortByLabel('spam')">Spam</button>
+    <button (click)="sortByLabel('trash')">Trash</button>
+  </div>
+
+  <div class="email-list">
+    <div *ngIf="loading">
+      <!-- Afficher le spinner ici pendant le chargement -->
+      <div class="spinner">Chargement en cours...</div>
```
Comment (Benja-san) [inline src/app/pages/google-email/google-email.component.html:23]: Is this loading spin will be used for all the app? if yes please use a component to inject in every necessary place

File: src/app/pages/google-email/google-email.component.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import { async, lastValueFrom } from 'rxjs';
+import { DeleteConfirmationComponent } from '../../components/delete-confirmation/delete-confirmation.component';
+import { MatDialog } from '@angular/material/dialog';
+
+@Component({
+  selector: 'app-google-email',
+  templateUrl: './google-email.component.html',
+  styleUrls: ['./google-email.component.scss'],
+})
+export class GoogleEmailComponent implements OnInit {
```
Comment (Benja-san) [inline src/app/pages/google-email/google-email.component.ts:13]: you do not need to implement onInit to use ngOnInit anymore

File: src/app/pages/google-email/google-email.component.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  mailDetails: {
+    labels: any;
```
Comment (Benja-san) [inline src/app/pages/google-email/google-email.component.ts:15]: avoid any type

File: src/app/pages/google-email/google-email.component.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+  ngOnInit() {
+    const user = localStorage.getItem('user');
+    if (user) {
+      this.userInfo = JSON.parse(user);
+    }
+    this.getEmails().then((r) => console.log(r));
+  }
+
+  async confirmDelete(mailDetail: any) {
```
Comment (Benja-san) [inline src/app/pages/google-email/google-email.component.ts:36]: what is a mailDetail composed of ? if you can catch it you can replace any type by a mailDetail Model or interface

File: src/app/pages/google-email/google-email.component.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+  async getEmails() {
+    if (!this.userInfo || this.loadingEmails) {
+      return;
+    }
+    this.loadingEmails = true;
+    try {
+      const userId = this.userInfo['info'].sub;
+      const messages = await lastValueFrom(this.googleApi.emails(userId));
+      console.log(messages);
```
Comment (Benja-san) [inline src/app/pages/google-email/google-email.component.ts:60]: suppress console logs

File: src/app/pages/google-email/google-email.component.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  }
+  formatDate(internalDate: string): Date {
+    return new Date(+internalDate);
+  }
+
+  sortByLabel(label: string) {
+    this.mailDetails = this.mailDetails.filter((mail) => mail.labels.includes(label));
+  }
+
+  extractBodyFromMail(mail: any): string {
```
Comment (Benja-san) [inline src/app/pages/google-email/google-email.component.ts:86]: what is the mail composed of ?

File: src/app/services/email.service.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import { Injectable } from '@angular/core';
+import { HttpClient, HttpHeaders } from '@angular/common/http';
+import { Observable, of } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class EmailService {
+  emails: any[] = [];
```
Comment (Benja-san) [inline src/app/services/email.service.ts:9]: Email[]

File: src/app/services/email.service.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  nextPageToken = '';
+
+  constructor(private http: HttpClient) {}
+
+  getEmails(): Observable<any[]> {
```
Comment (Benja-san) [inline src/app/services/email.service.ts:14]: remove any type

File: src/app/services/email.service.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    if (accessToken) {
+      const headers = new HttpHeaders().set('Authorization', `Bearer ${accessToken}`);
+
+      let url = 'https://www.googleapis.com/gmail/v1/users/me/messages';
+
+      if (this.nextPageToken) {
+        url += `?pageToken=${this.nextPageToken}`;
+      }
+
+      this.http.get<any>(url, { headers }).subscribe((res: any) => {
```
Comment (Benja-san) [inline src/app/services/email.service.ts:26]: remove any type by Email[] type

File: src/app/services/email.service.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  deleteEmail(emailId: string): void {
+    const accessToken = localStorage.getItem('access_token');
+
+    if (accessToken) {
+      const headers = new HttpHeaders().set('Authorization', `Bearer ${accessToken}`);
+
+      this.http
+        .delete<any>(`https://www.googleapis.com/gmail/v1/users/me/messages/${emailId}`, { headers })
+        .subscribe((res: any) => {
+          console.log('Email deleted successfully');
```
Comment (Benja-san) [inline src/app/services/email.service.ts:56]: toaster instead of console log

File: src/app/services/email.service.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        });
+    }
+  }
+
+  get selectedEmail(): Observable<any> | null {
```
Comment (Benja-san) [inline src/app/services/email.service.ts:61]: remove any type

File: src/app/services/google-api.service.ts [hunk 12]
```diff
@@ -38,12 +45,42 @@ export class GoogleApiService {
       });
     });
   }
+  emails(userId: string): Observable<any> {
+    return this.httpClient.get(`${this.gmail}/gmail/v1/users/${userId}/messages`, { headers: this.authHeader() });
+  }
 
-  isLoggedIn(): boolean {
-    return this.oAuthService.hasValidAccessToken();
+  getMail(userId: string, mailId: string): Observable<any> {
```
Comment (Benja-san) [inline src/app/services/google-api.service.ts:52]: remove any type

File: src/app/services/google-api.service.ts [hunk 13]
```diff
@@ -38,12 +45,42 @@ export class GoogleApiService {
...
+    return this.httpClient.get(`${this.gmail}/gmail/v1/users/${userId}/messages/${mailId}`, {
+      headers: this.authHeader(),
+    });
   }
 
+  sendMail(userId: string, mail: any): Observable<any> {
```
Comment (Benja-san) [inline src/app/services/google-api.service.ts:58]: remove any type

File: src/app/services/google-api.service.ts [hunk 14]
```diff
@@ -38,12 +45,42 @@ export class GoogleApiService {
...
+  }
+
+  async deleteEmail(userId: string, id: string): Promise<void> {
+    const accessToken = localStorage.getItem('access_token');
+    const headers = new HttpHeaders().set('Authorization', `Bearer ${accessToken}`);
+    await this.httpClient
+      .delete(`${this.gmail}/gmail/v1/users/${userId}/messages/${id}`, { headers })
+      .toPromise()
+      .then((res: any) => {
+        console.log('Email deleted successfully', res);
```
Comment (Benja-san) [inline src/app/services/google-api.service.ts:80]: replace console logs by toastr message

File: src/app/services/google-api.service.ts [hunk 15]
```diff
@@ -38,12 +45,42 @@ export class GoogleApiService {
...
+      })
+      .catch((error) => {
+        console.error('Error deleting email:', error);
```
Comment (Benja-san) [inline src/app/services/google-api.service.ts:83]: replace console logs by toastr message

**Review by Benja-san — CHANGES_REQUESTED** [review]

File: src/app/pages/google-email/google-email.component.html (hunk 1 above)
Comment (MatteiT) [reply src/app/pages/google-email/google-email.component.html:23]: no

File: src/app/pages/google-email/google-email.component.ts (hunk 5 above)
Comment (MatteiT) [reply src/app/pages/google-email/google-email.component.ts:60]: done

File: src/app/pages/google-email/google-email.component.ts (hunk 4 above)
Comment (MatteiT) [reply src/app/pages/google-email/google-email.component.ts]: done

File: src/app/services/email.service.ts (hunk 9 above)
Comment (MatteiT) [reply src/app/services/email.service.ts:26]: done

File: src/app/services/email.service.ts (hunk 10 above)
Comment (MatteiT) [reply src/app/services/email.service.ts:56]: done

File: src/app/services/email.service.ts (hunk 11 above)
Comment (MatteiT) [reply src/app/services/email.service.ts:61]: done

File: src/app/pages/google-email/google-email.component.ts [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+  async getEmails() {
+    if (!this.userInfo || this.loadingEmails) {
+      return;
+    }
+    this.loadingEmails = true;
+    try {
+      const userId = this.userInfo['info'].sub;
+      const messages = await lastValueFrom(this.googleApi.emails(userId));
+      console.log(messages);
+      const detailsPromises = messages.messages.map(async (message: any) => {
```
[the remark this answers is not in the document]
Comment (MatteiT) [reply src/app/pages/google-email/google-email.component.ts]: done

**Review by Benja-san — APPROVED** [review]

**Review by emericb — APPROVED** [review]

**Pull request merged**